src/allmydata/frontends/magic_folder.py

```diff
@@ -88,22 +106,109 @@ def __init__(self, client, upload_dircap, collective_dircap, local_dir, dbfile,
                              recursive=True)
 
-    def _check_db_file(self, childpath):
-        # returns True if the file must be uploaded.
+
+    def _should_download(self, path, remote_version):
+        """
+        _should_download returns a bool indicating whether or not a remote object should be downloaded.
+        We check the remote metadata version against our magic-folder db version number;
+        latest version wins.
+        """
+        v = self._db.get_local_file_version(path)
+        if v is None:
+            return True
+        else:
+            if v < remote_version:
+                return True
+            else:
+                return False
+
+    def _scan_remote(self, nickname, dirnode):
+        listing_d = dirnode.list()
+        self._download_scan_batch = {}
+        def scan_listing(listing_map):
+            for name in listing_map.keys():
+                file_node, metadata = listing_map[name]
+                if self._download_scan_batch.has_key(name):
+                    self._download_scan_batch[name] += [(name, file_node, metadata)]
+                else:
+                    self._download_scan_batch[name] = [(name, file_node, metadata)]
+        listing_d.addCallback(scan_listing)
+        return listing_d
+
+    def _scan_remote_collective(self):
+        if self._collective_dirnode is None:
+            return
+        upload_readonly_dircap = self._upload_dirnode.get_readonly_uri()
+        collective_dirmap_d = self._collective_dirnode.list()
+        def do_filter(result):
+            def not_mine(x):
+                return result[x][0].get_readonly_uri() != upload_readonly_dircap
+            others = filter(not_mine, result.keys())
+            return result, others
+        collective_dirmap_d.addCallback(do_filter)
+        def scan_collective(result):
+            d = defer.succeed(None)
+            collective_dirmap, others_list = result
+            for dir_name in others_list:
+                d.addCallback(lambda x, dir_name=dir_name: self._scan_remote(dir_name, collective_dirmap[dir_name][0]))
+                collective_dirmap_d.addCallback(self._filter_scan_batch)
+                collective_dirmap_d.addCallback(self._add_batch_to_download_queue)
+            return d
+        collective_dirmap_d.addCallback(scan_collective)
+        return collective_dirmap_d
+
+    def _add_batch_to_download_queue(self, result):
+        self._download_deque.extend(result)
+        self._download_pending.update(map(lambda x: x[0], result))
+
+    def _filter_scan_batch(self, result):
+        extension = []
+        for name in self._download_scan_batch.keys():
+            if name in self._download_pending:
+                continue
+            for item in self._download_scan_batch[name]:
+                (nickname, file_node, metadata) = item
+                if self._should_download(name, metadata['version']):
+                    extension += [(name, file_node, metadata)]
```

Comment:
Decide how this should work: do we want to scan and then filter download batches from each client individually, or scan all clients, then find the highest version number for each file and then filter?